README.md

```diff
@@ -816,6 +817,122 @@ You may need to do this for on a test-by-test basis as disabling class preload
 can cause problems in other use cases. For example, when you have library
 classes that require access to the `env` global.
 
+## Writing Testable Libraries
+
+We recommend the following best-practices for organizing pipeline code:
+
+* Keep complex logic in the `Jenkinsfile` to a minimum
+  - When possible, move complexity external scripts that the `Jenkinsfile` executes
+  - Move shared functionality to [pipeline libraries](https://www.jenkins.io/doc/book/pipeline/shared-libraries/)
+  - Likewise, any tricky Groovy logic that can't be easily moved to external scripts
+    should also be placed in pipeline libraries
+* In pipeline libraries, organize logic in classes under `src`
+  - Ideally, JenkinsPipelineUnit is used to test *only* these classes
+* Use the `vars` singletons to instantiate classes from `src`
+
+### On External Scripts
+
+In general, it's better to avoid having complex build logic inside of build pipelines.
+Although tools like JenkinsPipelineUnit are useful in testing pipelines, it's much easier
+to run build scripts locally (meaning, outside of a Jenkins environment). Languages like
+Python have much more sophisticated linting and testing tools than Groovy does.
```

Comment:
I know that CodeNarc can be used on regular Groovy code and Jenkinsfiles. Is that worth mentioning here?

Reply:
It is (and I do love that tool!), but it's nothing compared to the linters available in other languages. Still, I will mention it, though.